### DISCUSSION TITLE
Liquid soap advanced processing script

### DISCUSSION THREAD

**Discussion by sander816** [AzuraCast/AzuraCast#6923] [Q&A]
Does anyone have a good fm processing script for liquid soap, instead of stereotool?

It's taking too much cpu from my vps

**Comment by JimboJonesMGR** [comment]
Checking on a script. What kind of VPS do you use? Is it a paid account for example GoDaddy or other provider? They throttle the CPU sometimes due to network load. If its in house there are a couple of solutions. Try lowering the CPU usage/overhead by 2%. If that does not work keep lowering it 1% at a time until you get the desired results. 

UBUNTU has scripts that lower the CPU usage of a specific app or process.

If its Windows based you can lower the CPU usage the same way. However it may be as simple as setting the priority and affinity from the Task Manger. On Windows you can find .bat scripts that will allow you to lower the CPU usage of an app. You can add it to a schedule in to Task Manger. Set it to run every 5 minutes etc. 

What happens is the CPU just tries to run everything at 100% using all cores. You can define how many cores to use, the affinity and the priority. 

I had a similar problem with streaming apps. By adjusting the CPU throttle for the app or process to slightly lower than 100% you leave head room for the system. Think of it as getting too much fuel in to a carburator. The engine will over heat and have poor performance. Computers are the same way. Leaving some small amount of CU processing available for the system will make it run better.

Even if you get a script to do what you want you may still have problems if you do not adjust CPU throttling. 

With the UnReal Engine there are some issues exactly like this. I set the affinity, priority and core count just a little lower than the 100% capacity of the CPU. It actualy runs smoother and the CPU core temp is lower. The performance is smoother and I never have glitching or stalling when composing graphics and games.

Computers are so powerful now they can actualy over manipulate the code and software. Another issue is I/O data on your storage device. Hard drives and SSD can cause this effect. You can get scripts and pipe throttling software that can help tune u[ that issue as well. However, if you try everything else, it could be that the hard drive I/O, board bus or RAM is starting to fail. The CPU can be failing slightly or have issues with thread count. You can lower threads being used as well. 

Try to give the system a little head room where you can. It will make things go more smoothly.

 I hope this helps with your issue. Hit me back if you still have problems. I will brainstorm some more on a solution.

**Comment by sander816** [comment]
thats not what i mean:
in older versions i used this

`# Functions / Modules --------------------->

# Compressor Module / Limiter Module
compressor = ladspa.sc4(rms_peak = 0.5)
limiter = ladspa.tap_limiter(limit_level = -0.5)

# Single Band Compressor
def multiband_band(process, from, to, attack, release, ratio, threshold, gain)
	from_filter = filter.iir.eq.high(frequency = from,)
	to_filter = filter.iir.eq.low(frequency = to,)
	process_filtered = to_filter(from_filter(process))
	process_compressed = limiter(compressor(attack_time = attack, release_time = release, threshold_level = threshold, ratio=ratio, makeup_gain = gain, process_filtered))
	process_compressed
end

# Audio Input --------------------->

process = input

# Processing Blocks --------------------->

# Gate
process = ladspa.gate(process, threshold = -60.0, attack = 0.15, hold = 1.0, decay = 200.0, range = -25.0)

# Wideband AGC + Pre-Processing
process = normalize(target = 0., window = 0.03, gain_min = -16., gain_max = 0., process)
process = ladspa.sc4(rms_peak=0.3, attack_time = 0.5, release_time = 3., threshold_level = -36.0, ratio=1., makeup_gain = 6., process)

# Stereo Expander
process = ladspa.matrixspatialiser(width=16, process)

# Bass EQ
process = ladspa.tap_equalizer(band_1_freq=70., band_1_gain=4., process)
process = ladspa.tap_equalizer(band_1_freq=150., band_1_gain=-2., process)

# 5 Bands Compress/Limit
process = add(normalize=false,
  [
  multiband_band(process, 0., 200., 3.5, 30., 3., -10., 2.),
  multiband_band(process, 200., 1000., 2.25, 40., 2., -13., 2.),
  multiband_band(process, 1000., 3500., 2.25, 40., 3., -9., 2.),
  multiband_band(process, 3500., 6500., 2.25, 60., 2., -6., 1.5),
  multiband_band(process, 6500., 20000., 2.25, 70., 2., -4., 1.)
  ])

# 2 Bands Compress/Limit
process = add(normalize=false,
  [
  multiband_band(process, 0., 200., 10., 30., 2., -4., 0.),
  multiband_band(process, 200., 20000., 10., 40., 2., -2., -2.)
  ])

# De-esser
process = ladspa.tap_deesser(threshold_level=-5., frequency=6000., process)

# Final Limiter
process = limiter(process)

# Audio Output --------------------->

output = process`

**Comment by Moonbase59** [comment]
You could try what you get out of _master_me_, which should come bundled with AzuraCast. But for FM, Stereo Tool would normally be the best choice, since it can handle high sample rates, the stereo pilot and RDS data and other stuff like this. Plus, it makes calibration easy, like keeping modulation within legal range. Probably need to check your outputs using an external FM analyzer (or your F.M. Exciter).

All this functionality comes at a price, though (license, and high CPU usage). _master_me_ indeed uses _a lot less CPU_, because it’s intended for another audience and doesn’t offer all that features.

Probably depends much on your processing/broadcasting chain. I don’t have any practical experience with this, but I think AzuraCast can handle LADSPA plugins. It’s been 20 or more years since we did FM, so I’m not really up-to-date anymore…
